### PR TITLE
Contact Form: Add a button for activating the CRM JP Forms Extension

### PR DESCRIPTION
**Pull request by kbrown9**
#### Changes proposed in this Pull Request:
* Add a button to the CRM integration settings which allows the user to enable CRM's Jetpack Forms extension.
* Introduce two new REST API endpoints: one for obtaining information about the CRM plugin, and one for activating the CRM Jetpack Forms extension.
* Refactor the front-end code to improve clarity. These changes make the CRM integration code similar to the newsletter integration code. We might add plugin installation/activation buttons to CRM integration in the future, and this refactor should make those changes relatively straightforward.

#### Jetpack product discussion
- n/a

#### Does this pull request change what data or activity we track or use?
- no

##### Jetpack CRM not installed
1. Install, activate, and connect Jetpack.
2. Create a new post with the Jetpack Contact Form block.
3. Navigate to Posts and click the `edit` link for the post containing the contact form block.
4. Open the Forms block sidebar menu. For example, click on a form field, and then click `Form` in the breadcrumbs on the bottom of the page.
5. Open the CRM Integration drop-down. You should see text with a link to jetpackcrm.com
     Screenshot: https://d.pr/i/LHtMz0

##### Jetpack CRM installed but not activated
1. Install Jetpack CRM (do not activate it yet).
2. Navigate to Posts and click the `edit` link for the post containing the contact form block.
3. Open the Forms block sidebar menu. 
4. Open the CRM Integration drop-down. You should see text instructing you to activate Jetpack CRM.
    Screenshot: https://d.pr/i/oBKXlx

##### Jetpack CRM activated.
1. Navigate to Plugins and activate Jetpack CRM.
2. Navigate to Posts and click the `edit` link for the post containing the contact form block.
3. Open the Forms block sidebar menu. 
4. Open the CRM Integration drop-down. You should see a toggle switch that allows you to control the CRM integration.
    Screenshot: https://d.pr/i/nlwLvJ

##### CRM Jetpack Forms Extension deactivated.
1. Navigate to wp-admin -> Jetpack CRM -> Extensions. Scroll to the bottom of the page, and deactivate the Jetpack Forms extension.
2. Navigate to Posts and click the `edit` link for the post containing the contact form block.
3. Open the Forms block sidebar menu. 
4. Open the CRM Integration drop-down. You should see a button that allows you to enable the CRM Jetpack Forms extension.
    Screenshot: https://d.pr/i/DeZlUQ
5. Click the button to enable the extension.
6. You should see the toggle switch.

##### CRM Jetpack Forms Extension deactivated and the user is not an administrator.
Only administrators should be able to enable the Jetpack Forms Extension.
1. Create a non-admin user that has the `publish_posts` capability, for example, an editor.
2. Log in as the non-admin user.
3. Navigate to Posts and click the `edit` link for the post containing the contact form block.
4. Open the Forms block sidebar menu. 
5. Open the CRM Integration drop-down. You should see a message informing you that an admin needs to enable the extension.
    Screnshot: https://d.pr/i/6J5Vwz

#### Proposed changelog entry for your changes:
* Contact Form: add "Enable Jetpack Forms Extension" button to the CRM integration settings.

### CODE REVIEW

**Review by jeherve — COMMENTED** [review]
> Open the CRM Integration drop-down. You should see a button that allows you to enable the CRM Jetpack Forms extension.
> Screenshot: d.pr/i/DeZlUQ
> Click the button to enable the extension.
> You should see the toggle switch.

The status changes well in the editor, but the extension doesn't appear to get enabled; if I refresh the page I go back to the dialog asking me to enable the extension. I receive the right `{"code":"success"}` response from the `wp-json/jetpack/v4/jetpack_crm?_locale=user`, but the extension doesn't seem to be enabled.

**Comment by kbrown9** [comment]
> The status changes well in the editor, but the extension doesn't appear to get enabled; if I refresh the page I go back to the dialog asking me to enable the extension. I receive the right `{"code":"success"}` response from the `wp-json/jetpack/v4/jetpack_crm?_locale=user`, but the extension doesn't seem to be enabled.

That's unexpected! Here's a [video](https://d.pr/v/H89lho) that shows the extension is enabled for me.
The steps in the video are:

1. Deactivate the JPCRM Jetpack Forms extension.
2. Activate the JPCRM Jetpack Forms extension using the new button in the contact form sidebar menu.
3. Refresh the post editor page, and verify that the extension is enabled.
4. Verify that the extension is enabled on the JPCRM Extensions page.

I've been trying to think of reasons that it didn't work for you. When you click an `Activate` or `Deactivate` button on the JPCRM Extensions page, you're redirected to the extensions page with the `&zbsinstall=jetpackforms` parameter added to the url:

`/wp-admin/admin.php?page=zerobscrm-extensions&zbsinstall=jetpackforms&_wpnonce=...`

That parameter is used to toggle the extension on and off. So, if you refresh the page with that parameter in the URL, you will activate/deactivate the extension. Is it possible that you refreshed the extensions page with that parameter in the URL and disabled the extension?

**Comment by kbrown9** [comment]
I'm adding the 'Needs Review' label, but this shouldn't be merged until we've identified why the extension wasn't enabled in @jeherve's tests.

**Comment by jeherve** [comment]
> So, if you refresh the page with that parameter in the URL, you will activate/deactivate the extension. Is it possible that you refreshed the extensions page with that parameter in the URL and accidentally disabled the extension?

Bingo! That's exactly what happened; I kept the Extensions page opened in a separate tab to keep an eye on the extensions' status. I would refresh it after toggling the status in the block editor, to see if my changes had been applied.

> So, if you refresh the page with that parameter in the URL, you will activate/deactivate the extension.

Shouldn't the nonce stop this from happening? That seems like a bug in the plugin, no?

File: 3rd-party/class-jetpack-crm-data.php [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	 * sidebar menu.
+	 *
+	 * @return true|WP_Error Returns true if activation is success, else returns a WP_Error object.
+	 */
+	public function activate_crm_jetpackforms_extension() {
+		if ( function_exists( 'zeroBSCRM_extension_install_jetpackforms' ) ) {
+			return zeroBSCRM_extension_install_jetpackforms();
+		}
+
+		return new WP_Error( 'jp_forms_extension_activation_failed', 'The Jetpack Forms extension could not be activated.' );
```
Comment (jeherve) [inline 3rd-party/class-jetpack-crm-data.php:75]: Should we translate this? 

```suggestion
		return new WP_Error( 'jp_forms_extension_activation_failed', esc_html__( 'The Jetpack Forms extension could not be activated.', 'jetpack' );
```

**Review by jeherve — DISMISSED** [review]
This looks good for me now. I only have one minor nit-pick, but not a blocker, it can be addressed in a follow-up PR if needed.

**Comment by kbrown9** [comment]
> > So, if you refresh the page with that parameter in the URL, you will activate/deactivate the extension.
> 
> Shouldn't the nonce stop this from happening? That seems like a bug in the plugin, no?

Yes, it does seems like a bug. I'll open an issue for that.

File: 3rd-party/class-jetpack-crm-data.php (hunk 1 above)
Comment (kbrown9) [reply 3rd-party/class-jetpack-crm-data.php]: Good catch! I think this change can be added to this PR. See 8d0ad21.

File: 3rd-party/class-jetpack-crm-data.php [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	 * sidebar menu.
+	 *
+	 * @return true|WP_Error Returns true if activation is success, else returns a WP_Error object.
+	 */
+	public function activate_crm_jetpackforms_extension() {
+		if ( function_exists( 'zeroBSCRM_extension_install_jetpackforms' ) ) {
+			return zeroBSCRM_extension_install_jetpackforms();
+		}
+
+		return new WP_Error( 'jp_forms_extension_activation_failed', esc_html__( 'The Jetpack Forms extension could not be activated.', 'jetpack' );
```
Comment (jeherve) [inline 3rd-party/class-jetpack-crm-data.php:75]:
```suggestion
		return new WP_Error( 'jp_forms_extension_activation_failed', esc_html__( 'The Jetpack Forms extension could not be activated.', 'jetpack' ) );
```

**Review by jeherve — CHANGES_REQUESTED** [review]

**Review by jeherve — APPROVED** [review]

**Comment by jeherve** [comment]
r214737-wpcom

**Pull request merged**